### PR TITLE
Add pytorch 1.12.0 docker image

### PR DESCRIPTION
**Pull request by linden-li**
This PR adds support for PyTorch 1.12 to the build matrix. The new build matrix is:

- Python 3.9 + torch 1.12
- Python 3.8 + torch 1.11
- Python 3.7 + torch 1.10

TODO:

- [x] Merge https://github.com/mosaicml/composer/pull/1275

### CODE REVIEW

**Review by ravi-mosaicml — COMMENTED** [review]
The generate_build_matrix.py script will need to be changed. Run that locally then commit the result.

**Review by ravi-mosaicml — COMMENTED** [review]
Since pytorch 1.9 is now 4 versions outdated, let's drop support for it?

Thoughts about making the build matrix something like this?  CC @bandish-shah 

* Python 3.9 + torch 1.12
* Python 3.8 + torch 1.11
* Python 3.7 + torch 1.10

If we do drop testing for torch 1.9, can you update the `setup.py` and `meta.yaml` to reflect the minimum torch version of 1.10?

I think the torchvision version will also need to be updated for pytorch 1.12 -- usually they release them in parallel. https://pypi.org/project/torchvision/0.13.0/

**Comment by moinnadeem** [comment]
@ravi-mosaicml I wanted to flag a potential regression before we get far with PyTorch 1.12: https://mosaicml.atlassian.net/browse/CO-328

Basically, 1.12 disables TF32 matrix multiplications by default, so it might be slower. If we support 1.12, we should add a flag in Composer that checks your PyTorch version and re-enables TF32 (as long as we believe in it).

**Comment by abhi-mosaic** [comment]
> @ravi-mosaicml the proposed build matrix looks good to me.
> 
> We should get TF32 default change in asap as this will likely lead to performance regressions.
> 
> Should we get the soft cross entropy deprecation warning in for 0.8.1 and deprecate our implementation in 0.9.0?

I'm working on a precision refactor, and was gonna add`fp32` and `tf32` options within Composer. Does this solve the issue? Or you still want to make all our Docker images set `tf32` by default at the image level?

**Comment by hanlint** [comment]
This PR scope is to add the pytorch 1.12.0 docker image, lets not expand the scope to other work (e.g. deprecating pytorch versions) which is a composer-wide setting.

**Comment by ravi-mosaicml** [comment]
I would normally agree, but upgrading the image to pytorch 1.12 will introduce a regression with the tf32 / fp32 change for anyone using the latest mosaicml/pytorch image. Synced with @abhi-mosaic this morning; thinking of just adding in `tf32` without the precision refactor; then can merge this in. The additional refactoring of the precision enum will come later and will not block the pytorch 1.12 image.

File: docker/build_matrix.yaml
Comment (ravi-mosaicml) [inline docker/build_matrix.yaml:44]: I think the torchvision version will need an update -- they usually release a new torchvision version for each pytorch version.

File: docker/generate_build_matrix.py
Comment (ravi-mosaicml) [inline docker/generate_build_matrix.py:27]: See the comment re the build matrix -- can we bump python 3.7 to torch 1.10, and python 3.8 to torch 1.11?

File: docker/generate_build_matrix.py
Comment (ravi-mosaicml) [inline docker/generate_build_matrix.py:36]: My guess is we need 0.13.0 for torch 1.12?
```suggestion
        return '0.13.0'
```

**Review by ravi-mosaicml — COMMENTED** [review]
See comments: re pytorch + python + torchvisions versions. Will hold off on approving until the tf32 commit is merged.

File: docker/generate_build_matrix.py
Comment (linden-li) [reply docker/generate_build_matrix.py:36]: Yes, torchvision should be upgraded to 0.13.0

File: docker/generate_build_matrix.py
Comment (ravi-mosaicml) [inline docker/generate_build_matrix.py:37]:
```suggestion
    if pytorch_version == '1.10.2':
```

Also nit: can you you sort these checks in order? A bit confusing to jump from 1.11 to 1.12 and back to 1.12.

**Review by ravi-mosaicml — APPROVED** [review]
LGTM! Thanks for doing this and dealing with all of the bugs that came up.

**Pull request merged**
